Features/task 1

milan009 opened the pull request.

Commit:
KA-183 Add ListItem model

Commit:
KA-183 Install Microsoft.AspNet.WebApi.Versioning NuGet package

Commit:
KA-183 Add Items controller, implement URL versioning

Commit:
KA-183 Implement the HTTP methods, revision versioning

Commit:
KA-183 Refactor the controller to better reflect the task

Commit:
KA-183 Install NUnit package, add dummy test

Commit:
KA-183 Extract the dummy data into constants class

Commit:
KA-183 Add equality comparer, implement unit tests

Commit:
KA-183 Cleanup of code

Commit:
KA-183 Removed unused references

Commit:
Merge branch 'features/task-1' of https://github.com/milan009/kentico-onboarding-cs into features/task-1

PetrSvirak commented on ListApp/ListApp.Api/Controllers/ItemsController.cs:
```diff
@@ -0,0 +1,43 @@
+
+namespace ListApp.Api.Controllers
+{
+    namespace V1
```
Always stick with `namespaces` by folders. If you need a `.V1` namespace, create a `V1` folder within the `Controllers` folders. Only reason to declare `namespace` like this, would be a plan to place another declaration later on (e.g. `.V2`). This would though result in a quite long file with 2 classes in it.

PetrSvirak commented on ListApp/ListApp.Api/Controllers/ItemsController.cs:
```diff
@@ -0,0 +1,43 @@
+        [RoutePrefix("api/v{version:apiVersion}/items")]
+        public class ItemsController : ApiController
+        {
+            #region HTTP verbs implementations
```
Do not use regions. It almost always shows that your code is badly structured (typically it points to an SRP violation. It also does not bring any extra value since there is 1) a single region and 2) content of the region is pretty short.

PetrSvirak commented on ListApp/ListApp.Api/Controllers/ItemsController.cs:
```diff
@@ -0,0 +1,43 @@
+                => await Task.FromResult(Ok(Constants.MockListItems));
+
+            [Route("{id}")]
+            public async Task<IHttpActionResult> GetAsync([FromUri]Guid id) 
```
I encourage you to put a space in between attribute and type.

PetrSvirak commented on ListApp/ListApp.Api/Controllers/ItemsController.cs:
```diff
@@ -0,0 +1,43 @@
+
+            [Route("{id}")]
+            public async Task<IHttpActionResult> PutAsync([FromUri]Guid id, [FromBody]ListItem newItem) 
+                => await Task.FromResult(Created($"/items/{id}", newItem));
```
Applies on both `POST` and `PUT`:
* Use fixed existing items instead of returning data posted/put by user (since you do not validate request, you cannot be sure the response (based on `newItem`) is correct)
* This header is actually wrong. Look at the location in `post-man` – it shows `location →/items/580d51d9-3ebe-4aac-a361-45c21498e5d4` while it should cleary return something like `/api/v1/items/00000000-0000-0000-0000-000000000002` – use `(base.)Url.Route` instead of the string.

PetrSvirak commented on ListApp/ListApp.Api/Utils/Constants.cs:
```diff
@@ -0,0 +1,23 @@
+    {
+        public static Guid Guid1 = Guid.Parse("00000000-0000-0000-0000-000000000000");
+        public static Guid Guid2 = Guid.Parse("00000000-0000-0000-0000-000000000001");
+        public static Guid Guid3 = Guid.Parse("00000000-0000-0000-0000-000000000002");
```
* since the GUIDs seems to be anywhere but within the class, definitely make them private.
* use `readonly` since you don't want these to change during execution (you are looking for closest equivalent to `const` here and R# should actually suggest this to you)

PetrSvirak commented on ListApp/ListApp.Api/Utils/Constants.cs:
```diff
@@ -0,0 +1,23 @@
+        public static Guid Guid3 = Guid.Parse("00000000-0000-0000-0000-000000000002");
+        public static Guid NonExistingItemGuid = Guid.Parse("00000000-0000-0000-0000-000000000003");
+
+        public static IEnumerable<ListItem> MockListItems => new List<ListItem>
```
Using an array would server better purpose here (since it is not permanent API and you want to access it by index (`MockListItems[0]` vs `MockListItems.ElementAt(0)`))

PetrSvirak commented on ListApp/ListApp.Api/Utils/Constants.cs:
```diff
@@ -0,0 +1,23 @@
+        {
+            new ListItem {Id = Guid1, Text = "Stretch correctly"},
+            new ListItem {Id = Guid2, Text = "Make a coffey"},
+            new ListItem {Id = Guid3, Text = "Take over the world"}
```
Consider making the while `ListItem`s `private`, `static` and `readonly`, since you needn't change them anyway.

PetrSvirak commented on ListApp/ListApp.Api/App_Start/WebApiConfig.cs:
```diff
@@ -11,12 +10,22 @@ public static void Register(HttpConfiguration config)
-            config.MapHttpAttributeRoutes();
+            config.MapHttpAttributeRoutes(constraintResolver);
 
             config.Routes.MapHttpRoute(
```
One does not need default route when using `attribute routing` - it might hide some errors (like missing route attributes) - generally, use either one or the other.

PetrSvirak commented on ListApp/ListApp.Api/App_Start/WebApiConfig.cs:
```diff
@@ -11,12 +10,22 @@ public static void Register(HttpConfiguration config)
+                }
+            };
+
+            config.AddApiVersioning();
```
Really nice 👍 Both the resolver and config.

PetrSvirak commented on ListApp/ListApp.Api/App_Start/WebApiConfig.cs:
```diff
@@ -11,12 +10,22 @@ public static void Register(HttpConfiguration config)
         {
             // Web API configuration and services
 
+            var constraintResolver = new DefaultInlineConstraintResolver
```
Move resolver construction into a separate method - seems like a bit different level of abstraction than the rest of the class.

PetrSvirak commented on ListApp/ListApp.Api.Tests/Utils/ListItemEqualityComparer.cs:
```diff
@@ -0,0 +1,25 @@
+﻿using System.Collections.Generic;
+using ListApp.Api.Models;
+
+namespace ListApp.Api.Tests.Utils
```
Right now, it is a folder with `Comparers`, right? Why not name it so?

PetrSvirak commented on ListApp/ListApp.Api.Tests/Utils/ListItemEqualityComparer.cs:
```diff
@@ -0,0 +1,25 @@
+    /// It does NOT implement its own GetHashCode() method - use ONLY
+    /// for comparing two <see cref="ListItem"/> models!
+    /// </summary>
+    public class ListItemEqualityComparer : IEqualityComparer<ListItem>
```
That is all nice until you realize how you use it
* every time, in every test you create a new instance of the very comparer that consists of 2 pure functions → you should only use one static instance of the class (it is actually perfect candidate for a singleton)
* to move the singleton pattern a bit further, you will end-up with repeating `….using(…StaticComparerInstance)` → to make things more readable, why not introduce an extension method (call, e.g. `usingListItemComparer()`) that would be easy to find by IntelliSense and describe what's going on perfectly (→ not the file would best fit into a folder called `Extensions`)

PetrSvirak reviewed (changes requested):
Nice and simple. Great. There are some minor changes suggested below and there are some things to do with tests, but otherwise, I like your solution.

### The tests
* Do not use `Constants.MockListItems` here, since you actually do not want to rely on any invariant of tested code (here, the duplicity is welcomed)
* Respect the `AAA` in the tests, it renders them more navigable. If there is block that is too thick or repetitive, feel free to introduce some private methods (be careful not to overcomplicate the things though, there will always be just a bunch of tests here in this particular project, so do not start creating your own framework (or big framework extension))
* In connection with the previous bullet, use multiple `asserts` in one test. It is OK when either each `assertion` cannot be determined (or it does not make sense to `assert` it) when all previous `assertion`s failed or when you "trigger" all the `assert`s in the test everytime – the later would require a framework extension again that would be overkill.  Anyway, in this project, merge the assertion for all given calls (test that only differ in assertion part).
* The casting you do is too much connected with implementation and does not clearly point out what is actually tested (you are interested in API responses, not the framework implementation). So, use `wait receivedResponse.ExecuteAsync(…)` and then `response.Status` and `response.TryGetContentValue(…)` to obtain data from the API response.

Commit:
KA-183 Remake the comparer into a singleton, refactor code

An extension method for the NUnit "EqualConstraint" type called "UsingListItemComparer" has been added for the sake of increasing code readability.

Commit:
KA-183 Remove default routing, refactor the resolver initialization

Commit:
KA-183 Change the mock list into an array, make the guids private

Commit:
KA-183 Fix headers for PUT and POST, return predefined item

Commit:
KA-183 Add NSubstitute package

Commit:
KA-183 Combine tests, use "ExecuteAsync()" instead of casting

Also, stopped using the MockListItems in tests and removed regions.

Commit:
KA-183 Removed unused directives

Commit:
KA-183 Add spaces between attribute and type

Commit:
KA-183 Modify MockListItems to return a readonly array

Formerly, a new instance of this constant array had been created to avoid modification. This approach is no longer needed.

Commit:
KA-183 Refactor the constants - items are now private

PetrSvirak commented on ListApp/ListApp.Api.Tests/ItemControllerTests.cs:
```diff
@@ -25,105 +26,96 @@ public class ItemsControllerTests
-            var receivedItems = ((OkNegotiatedContentResult<IEnumerable<ListItem>>)receivedResponse).Content;
-            Assert.That(receivedItems, Is.EqualTo(Constants.MockListItems).Using(new ListItemEqualityComparer()));
+            Assert.AreEqual(expectedResponseCode, responseMessage.StatusCode);
+            Assert.IsTrue(responseMessage.TryGetContentValue(out IEnumerable<ListItem> responseItems));
```
One actually doesn't mind this. If the result is false, the actual item will not equal expected item. Moreover, the method call seems to fit better into act part (as you are aiming for actual item and not whether it is retrievable).

PetrSvirak commented on ListApp/ListApp.Api.Tests/ItemControllerTests.cs:
```diff
@@ -25,105 +26,96 @@ public class ItemsControllerTests
+            const HttpStatusCode expectedResponseCode = HttpStatusCode.OK;
+            var expectedItem = new ListItem
+            {
+                Id = Guid.Parse("00000000-0000-0000-0000-000000000000"),
```
There is this kinda special `Guid.Empty` 😛

PetrSvirak commented on ListApp/ListApp.Api.Tests/ItemControllerTests.cs:
```diff
@@ -25,105 +26,96 @@ public class ItemsControllerTests
 
         [Test]
-        public async Task Get_WithAnyId_ResponseIsOfCorrectType()
+        public async Task Get_WithAnyId_ResponseIsOfCorrectTypeAndReturnsFirtsItem()
```
Overlooked it before. It is actually not "any" ID, but rather "existing" ID, isn't it?

PetrSvirak commented on ListApp/ListApp.Api.Tests/ItemControllerTests.cs:
```diff
@@ -25,105 +26,96 @@ public class ItemsControllerTests
         {
-            var receivedResponse = await _itemsController.PutAsync(PostedItemGuid, PostedItem);
-            Assert.IsInstanceOf<CreatedNegotiatedContentResult<ListItem>>(receivedResponse);
+            var expectedLocation = _itemsController.Url.Request.RequestUri + "/00000000-0000-0000-0000-000000000003";
```
There is this `PostedItemGuid` that has absolutely same value. Avoid repetition within the tests (it is correct to copy items from your application to tests, especially since the application will evolve and get rid of the constants shortly, however, within the tests themselves, same (or even more strict) rules as for any other code apply)

PetrSvirak commented on ListApp/ListApp.Api/App_Start/WebApiConfig.cs:
```diff
@@ -9,25 +9,21 @@ public static class WebApiConfig
+            config.MapHttpAttributeRoutes(InitializeConstraintResolver());
+        }
+
+        private static IInlineConstraintResolver InitializeConstraintResolver()
```
Nice, just make it an expression bodied function.

PetrSvirak commented on ListApp/ListApp.Api.Tests/Comparers/ListItemEqualityComparer.cs:
```diff
@@ -10,6 +10,11 @@ namespace ListApp.Api.Tests.Utils
     /// </summary>
     public class ListItemEqualityComparer : IEqualityComparer<ListItem>
```
Move this class into the `NUnitExtensions` since this is the great example of correct private class usage.

PetrSvirak commented on ListApp/ListApp.Api.Tests/Extensions/NUnitExtensions.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace ListApp.Api.Tests.Extensions
+{
+    public static class NUnitExtensions
```
This would end up in quite some mess if you concentrated all NUnit extensions into one file. It is (IMHO good) custom to name extension classes by the type they extending → EqualConstraintExtensions (or if there were too much code, extensions can be named by features they are adding).

PetrSvirak commented on ListApp/ListApp.Api.Tests/Extensions/NUnitExtensions.cs:
```diff
@@ -0,0 +1,13 @@
+{
+    public static class NUnitExtensions
+    {
+        public static EqualConstraint UsingListItemComparer(this EqualConstraint constr)
```
Make it expression bodied function...

PetrSvirak commented on ListApp/ListApp.Api/Controllers/V1/ItemsController.cs:
```diff
@@ -0,0 +1,36 @@
+
+        [Route]
+        public async Task<IHttpActionResult> PostAsync([FromBody] ListItem newItem) 
+            => await Task.FromResult(Created(Url.Request.RequestUri + $"/{Constants.NonExistingItemGuid}", Constants.CreatedListItem));
```
It is not technically wrong to use `NonExistingItemGuid` here, however, consider using `Constants.CreatedListItem.Id` to better show the purpose.

PetrSvirak reviewed (changes requested):
I like most of the changes you did. Good job. There just a few minor details and then this quite big one:

You lost history when moving `ItemsController`. This is caused by doing too many changes all along with moving (renaming) the file around. You basically need to split this into two separate commits (one with changes, the other with file movement) rebasing the branch - ask Vlastik for assistance, he has the experience.

Commit:
KA-183 Reuse GUID constants

Commit:
KA-183 Include the comparer as a private class of the extension

Commit:
KA-183 Rename the NUnitExtensions

Commit:
KA-183 Use expression bodied functions

Commit:
KA-183 Add spaces to attributes vol. II

Commit:
KA-183 Move ItemsController to V1 folder

Commit:
KA-183 Fix namespace in ItemsController
